### ISSUE TITLE
[HOLD for payment 2025-02-07] [$250] Wallet - "It's not here" error is briefly shown when deactivating a virtual card

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#53996]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team
**Slack conversation** (hyperlinked to channel name):

## Action Performed:
Precondition: Use a Gmail member account that has a single assigned virtual Expensify Card.

1. Navigate to https://staging.new.expensify.com/
2. Log in
3. Navigate to Settings - Wallet
4. Click on the card under "Assigned cards"
5. Click on "Report virtual card fraud"
6. Click on the "Deactivate card" button
7. Input the magic code

## Expected Result:
I should be getting a proper error message at the bottom of the RHP.

## Actual Result:
"It's not here" error is briefly shown when deactivating a virtual card.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [z] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/6c7a3d06-05da-471e-8073-333cce30327e

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021867191340390551829</li>
        <li>Upwork Job ID: 1867191340390551829</li>
        <li>Last Price Increase: 2024-12-12</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by beodw** [comment]
Video of reproducing bug is not playing for me.

**Comment by mountiny** [comment]
@koko57 would you want to work on this?

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- "It's not here" error is briefly shown when deactivating a virtual card.

### What is the root cause of that problem?
- After step 7, we call:

https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L67

It will set the card data to null:

https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/libs/actions/Card.ts#L55

and then call:

https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L109

which will set the card data to ```{isLoading: false}```.

- As a result, when the API is still processing, the card data is ```{isLoading: false}```.

- When the API is finished, we have a logic to navigate the user back to the previous screen once the API is finished:

https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L59

the navigation action needs time to process. Meanwhile, we display the not found page since BE has set the card data to null:

https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L84-L86


- It is why the "It's not here" screen is briefly shown when deactivating a virtual card.
### What changes do you think we should make in order to solve the problem?
- We can introduce a state ```[isNavigating, setIsNavigating] = useState(false)``` and when ```isNavigating``` is ```true```, display the fullscreen loading indicator:

```
    if (isNavigating) {
        return <FullScreenLoadingIndicator />;
    }
```

- The ```isNavigating``` will be set to ```true``` after line:

https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L59

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?


### What alternative solutions did you explore? (Optional)

- Instead if display the loading indicator, we can update two conditions:

https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L84
to:

```
    if (isEmptyObject(virtualCard) && !isNavigating) {
```


and https://github.com/Expensify/App/blob/258e2be76bc20d038be2a24801856ec82b7a8581/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L100
to:

```
                    isLoading={formData?.isLoading || isNavigating}
```

**Comment by burczu** [comment]
Hi! I'm Bartek from Callstack - expert contributor group. I’d like to work on this issue (if you decide to decline the above proposal, of course).

**Comment by alitoshmatov** [comment]
@truph01 Thank you for your proposal. In your RCA you are mentioning that `isLoading` will be set to `false` even if request hasn't been completed, so I think this is a valid root cause. But I think your solution is ignoring this root cause and applying workaround

**Comment by koko57** [comment]
@burczu actually I'm working on a similar issue that probably have a similar root cause. If I may I will continue on these both 🙏🏻

**Comment by alitoshmatov** [comment]
@truph01 **Correction**: Okay I think `isLoading` is not the related at all, it is working just fine. Loading is shown when API is being processed and when api resolves successfully, the page starts navigating, while it is navigating briefly not found page is shown.

**Comment by alitoshmatov** [comment]
Based on C+ doc, since existing proposal is not selected we can assign @burczu to this issue and wait for their proposal

**Comment by alitoshmatov** [comment]
@truph01 Did you tested your alternative solution? 
https://github.com/Expensify/App/blob/919c59d0afe4ed7122d9bd714d93254ca4849060/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L109
We have a usage of `virtualCard` here. To be fair this small issue and could be fixed easily. But I am still in doubt about `isNavigating` state, I haven't seen similar use case in other places

**Comment by mountiny** [comment]
@burczu please go ahead

**Comment by burczu** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The "It's not here" 404 page is shown for a short period of time after successfully deactivating a virtual card and before redirecting back.

### What is the root cause of that problem?

To root cause of the problem is that we rely on existence of the virtual card inside the list of cards when deciding if we need to show the "It's not there" page or not:

https://github.com/Expensify/App/blob/c45f8342d36f09fa2d185650fa8d832d920737b9/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L84-L86

where the `virtualCard` variable is defined here (the `cardID` is taken from the route):

https://github.com/Expensify/App/blob/c45f8342d36f09fa2d185650fa8d832d920737b9/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L41

The problem is, that when we call the `ReportVirtualExpensifyCardFraud` API action, it sets the `cardList?.[cardID]` to `null` in response (we even set it like that optimistically). This causes the short period of time between the API response and starting the redirect back to the card page, where the condition to show the 404 page is fulfilled.

### What changes do you think we should make in order to solve the problem?


To prevent this, we could store the copy of the `cardList?.[cardID]` initial value inside the `ref`, and rely on it in the condition:

```javascript
const virtualCard = cardList?.[cardID];
const virtualCardRef = useRef(virtualCard);

...


if (isEmptyObject(virtualCardRef?.current)) {
    return <NotFoundPage />;
}
```

Thanks to these changes, we prevent showing the 404 error page after the API call is successful, even that the card is already removed from the list of cards. And it will still work for the original purpose of this part of code - if we open the page with incorrect `cardID` the 404 page will be presented.

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?


### What alternative solutions did you explore? (Optional)

n/a

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by mountiny** [comment]
@alitoshmatov thoights?

**Comment by mountiny** [comment]
@burczu does this also mean that if i am looking at the page as admin 1 and admin 2 deactivates this card, the pusher event will come in but i as admin 1 will still see the card as if it was not deleted?

**Comment by burczu** [comment]
@mountiny Hmm... I didn't consider this scenario... I'll test it today and if it is the case, I'll work on addressing it.

**Comment by mountiny** [comment]
I am not sure if its critical, these dynamic bugs are hard to handle, but feels like this might be the case with the prosed changes

**Comment by alitoshmatov** [comment]
I am not sure why but I don't think this approach is correct, we shouldn't have a copy of data just for the sake of solving slow navigation. If we needed it, then we should have set up our logic so that data be deleted after navigation happens. 

Also if data is set to null optimistically why wait for api request, based on the code shouldn't `NotFoundPage` be shown immediately.

https://github.com/Expensify/App/blob/0bf6d8cb0ae8fcd43819edf998131d9bc3cc7f68/src/pages/settings/Wallet/ReportVirtualCardFraudPage.tsx#L84-L86

**Comment by burczu** [comment]
@alitoshmatov Even without setting it optimistically, the API response sets it to `null` which is too early. At first I was trying to rely on the `isLoading` property and postponing changing it to `false` on the page we are navigating to (by adding additional query string parameter to the URL) but without success.

**Comment by burczu** [comment]
@mountiny I think the scenario you was worrying about is impossible - from what I see, every user can report fraud only of its own cards in the wallet, so it's impossible that another admin can see the same report fraud page at the same time. But maybe I'm missing something?

**Comment by burczu** [comment]
@alitoshmatov I've tired removing optimistically clearing card data in the list of cards but it does not fix the issue - the 404 page still shows up, but for shorter period.

I've also gave a try once again the solution I've described in my [previous comment](https://github.com/Expensify/App/issues/53996#issuecomment-2556374667) but unfortunately without success - the navigation finishes earlier than the animation so the 404 page is visible for some short time.

I think, as we can't react to the end of the animation (at least I couldn't find the way to do it), the only solution must be kinda hack-ish - I agree it's not perfect. If we don't like storing the copy of the data inside the `ref`, we could think of storing it in Onyx, under the `REPORT_VIRTUAL_CARD_FRAUD` key. What do you think?

**Comment by mountiny** [comment]
I kind feel like this flow is a bit weird

When i actually want to report fraud on my card, i am probably worried about abuse of the card. I would like to get clear feedback the card was successfully terminated and it could not be abused by anyone.

@JmillsExpensify @Expensify/design what do you this about showing some success screen after the card is deactivated? This would provide clear feedback to the user and fix this bug as well. 

We could also further improve the success screen by directly adding not only "ok" button, but a CTA to create a new card (if you are admin)

What do you think?

**Comment by mountiny** [comment]
Adding the label here again. Would love to get design perspective on improving the UX of this flow as raised in [here](https://github.com/Expensify/App/issues/53996#issuecomment-2558526713) in high level

**Comment by trjExpensify** [comment]
I like some form of confirmation. Using a success screen probably makes the most sense. I thought about that card art updating to show a deactivated state perhaps?

**Comment by shawnborton** [comment]
I like the idea as well! Maybe something like this? cc @Expensify/design 
![CleanShot 2025-01-10 at 10 21 34@2x](https://github.com/user-attachments/assets/b891acbe-c38a-48cf-8202-3a1c6a3bf629)

This way we can use this opportunity to have some copy that explains that when they go back to the card view, the new card details should be there. Right now they aren't populating but I am pretty sure we're fixing that in a different issue.

**Comment by burczu** [comment]
Ok, so as a summary: instead of redirecting to the card list after successful card deactivation, we want to just show the above screen - is that right?
I'll update the proposal, to fulfill this requirement (not sure if I'll manage to do it today as I'm focused on other work right now).

**Comment by shawnborton** [comment]
Yup, but let's just make sure others have a chance to weigh in to make sure they are on board. And cc @joekaufmanexpensify since this is card related and it might be of interest to you.

**Comment by trjExpensify** [comment]
Gosh, I love that fraud guy so much!

**Comment by dannymcclain** [comment]
Absolutely love it! Fraud guy for the win 😂 Also love the little explanation below to make the experience more clear.

>(Also sorry Danny, I know this was assigned to you but I really wanted to take our little fraudster illustration for a spin!)

I stole one from Jon a couple weeks ago, we're all just part of the circle of ~life~ design!

**Comment by dannymcclain** [comment]
Get that creeeeeddddiiitttttt

**Comment by shawnborton** [comment]
Haha thank u for the GitCoin sir!

**Comment by dubielzyk-expensify** [comment]
> I stole one from Jon a couple weeks ago, we're all just part of the circle of life design!

My steal is up next. So glad you got that illustration in 😄

**Comment by burczu** [comment]
Ok then, so who should work on this one? ;)

**Comment by mountiny** [comment]
@burczu would be great if you could work on it. Do you have everything you need?

**Comment by burczu** [comment]
@mountiny Yes, I think so.

**Comment by mountiny** [comment]
@burczu Let us know about your next steps and ETA. Thank you!

**Comment by burczu** [comment]
@mountiny I'm close to provide the solution - the proposal should be updated today.

**Comment by burczu** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/53996#issuecomment-2551250077)

I've also added the [draft PR here](https://github.com/Expensify/App/pull/55278), where you can test these changes.

**Comment by mountiny** [comment]
@alitoshmatov can you please have a look

**Comment by shawnborton** [comment]
I thought that after you report card fraud, you automatically re-generate a new virtual card. Is that not the case? If so, why isn't this page populated with the correct new information?

**Comment by alitoshmatov** [comment]
New virtual card is regenerated, the proposed solution is redirecting to the old card. We can redirect to new card

**Comment by shawnborton** [comment]
I would think we should redirect to the new card with a message. cc @joekaufmanexpensify but we had already discussed this here: https://github.com/Expensify/App/issues/52953#issuecomment-2563838963

And it seems like the PR for it is here: https://github.com/Expensify/App/pull/55163

**Comment by alitoshmatov** [comment]
Looks like this PR is working on the same place as we are which might cause some confusing. I think we should align with them and probably single PR would solve both of the issues. Not sure who should hold their issue

**Comment by shawnborton** [comment]
Perhaps we wait until the other PR is merged and then we can revisit and see if we are still getting the issue? @narefyev91 can you confirm if your PR would fix this bug or not?

**Comment by narefyev91** [comment]
> Perhaps we wait until the other PR is merged and then we can revisit and see if we are still getting the issue? [@narefyev91](https://github.com/narefyev91) can you confirm if your PR would fix this bug or not?

Yeah - based on my testing - not found page is not shown any more

**Comment by shawnborton** [comment]
Ah okay - so @JmillsExpensify @mountiny maybe we close this then?

**Comment by shawnborton** [comment]
Sounds good, that works for me - thanks for the gut check here.

**Comment by burczu** [comment]
@mountiny @alitoshmatov I've synced with the changes from [this PR](https://github.com/Expensify/App/pull/55163), and it looks like it does not interfere with my changes so I'm keeping the proposal as it is. Especially we need to keep the `cardID` stored in the `ONYXKEYS.CARD_LIST` key solution because the use of `usePrevious` hook from the other PR does not work: in our case cause we need to show additional modal with confirmation and it causes extra re-render so the previous `virtualCard` value is lost.

Regarding this comment: https://github.com/Expensify/App/issues/53996#issuecomment-2602582148 this is fixed by the other PR so once we get back to this page, it is correctly populated with new card details.

That being said, can I proceed with the PR now?

**Comment by mountiny** [comment]
Thanks for the discussion, I _believe_ you can move ahead now if I understood everything right

**Comment by burczu** [comment]
[The PR](https://github.com/Expensify/App/pull/55278) is now ready for review.

**Comment by mountiny** [comment]
This will be ready to pay out

**Comment by alitoshmatov** [comment]
Let's also wait for waiting period of regression issues:
1. https://github.com/Expensify/App/issues/56023 - 10th Feb
2. https://github.com/Expensify/App/issues/56026 - 13th Feb

**Comment by alitoshmatov** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by garrettmknight** [comment]
**Payment Summary:**
-  @alitoshmatov owed $250 via NewDot

@alitoshmatov request whenever you're ready.

**Comment by alitoshmatov** [comment]
Requested payment in ND

**Comment by JmillsExpensify** [comment]
$250 approved for @alitoshmatov
